pkg/package/template.go

```diff
@@ -0,0 +1,205 @@
+	VariableSchemas map[string]interface{} `yaml:"variables,omitempty"`
+}
+
+func Template(name, file string) error {
+
+	body, err := os.ReadFile(file)
+	if err != nil {
+		return err
+	}
+
+	var t templateSpec
+
+	if err = yaml.Unmarshal(body, &t); err != nil {
+		return err
+	}
+
+	pkgs := make([]Package, len(t.Packages))
+	for i, p := range t.Packages {
+		pkg, err := LoadPackage(p) // ensures pkg is in cache
+		if err != nil {
+			return err
+		}
+		pkgs[i] = pkg
+	}
+
+	manifest, err := mergePackages(name, pkgs)
+	if err != nil {
+		return err
+	}
+
+	manifest.Description = t.Description
+
+	buf, err := yaml.Marshal(manifest)
+	if err != nil {
+		return err
+	}
+
+	err = ValidateManifest(buf)
+	if err != nil {
+		return err
+	}
+
+	err = os.WriteFile(filepath.Join(name, "manifest.yaml"), buf, 0664)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func mergePackages(name string, pkgs []Package) (TemplateManifest, error) {
+	t := TemplateManifest{
+		Name:            name,
+		Overlay:         map[string]string{},
+		VariableSchemas: map[string]interface{}{},
+	}
+	var srcs = map[string]string{}
+	for _, pkg := range pkgs {
+		buf, err := ioutil.ReadFile(path.Join(pkg.RootPath, "manifest.yaml"))
+		if err != nil {
+			return t, err
+		}
+
+		yml := struct {
+			VariableSchemas map[string]interface{} `yaml:"variables,omitempty"`
+		}{
+			VariableSchemas: map[string]interface{}{},
+		}
+
+		err = yaml.Unmarshal(buf, &yml)
+		if err != nil {
+			return t, err
+		}
+
+		for _, c := range pkg.Commands {
+			if c.Module != "" {
+				c.Module = fmt.Sprintf("%s-%s", pkg.Name, c.Module)
+			}
+			t.Commands = append(t.Commands, c)
+		}
+		for k, v := range pkg.Overlay {
+			t.Overlay[k] = v
+		}
+		for k, v := range yml.VariableSchemas {
+			if schema, ok := t.VariableSchemas[k]; ok {
+				if !reflect.DeepEqual(schema, v) {
+					logrus.Fatalf("Variable \"%s\" has conflicting schemas:\n%s:\t%v\n%s:\t%v", k, srcs[k], schema, pkg.Name, v)
+				}
+				continue
+			}
+			t.VariableSchemas[k] = v
+			srcs[k] = pkg.Name
+		}
+
+		logrus.Infof("Copying modules from %s", pkg.Name)
+
+		err = copyTerraform(name, pkg)
+		if err != nil {
+			return t, err
+		}
+
+		logrus.Infof("Copying overlay from %s", pkg.Name)
+
+		err = copyOverlay(name, pkg)
+		if err != nil {
+			return t, err
+		}
+
+	}
+	return t, nil
+}
+
+func copyTerraform(root string, pkg Package) error {
+	return copyFiles(path.Join(root, "terraform", pkg.Name), pkg.TerraformRootPath(), "-", pkg)
+}
+
+func copyOverlay(root string, pkg Package) error {
+	return copyFiles(path.Join(root, "overlay"), pkg.OverlayPath(), "/", pkg)
+}
+
+func copyFiles(root, dir, sep string, pkg Package) error {
+	_, err := os.Stat(dir)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+	err = filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if path == dir {
+			return nil
+		}
+
+		orig := strings.TrimPrefix(path[len(dir):], "/")
```

Comment:
there is no need to trim the prefix here POSIX allows for multiple slashes in a path.